### PR TITLE
http POST support

### PR DESCRIPTION
**Pull request by fabricedesre**
@Arqu I could get some help with something in the http handler: the new `put_handler` itself compiles fine but anytime I use the unixfs builder in that function (uncommenting any of the `let file = create_file_from_xxx` line) I get a trait bound error from the route setup that I don't understand:

```rust
error[E0277]: the trait bound `fn(Extension<Arc<core::State<T>>>, http::Request<hyper::Body>) -> impl futures::Future<Output = Result<response::GatewayResponse, GatewayError>> {put_handler::<T>}: Handler<_, _>` is not satisfied
   --> iroh-gateway/src/handlers.rs:71:33
    |
71  |         .route("/:scheme/", put(put_handler::<T>))
    |                             --- ^^^^^^^^^^^^^^^^ the trait `Handler<_, _>` is not implemented for `fn(Extension<Arc<core::State<T>>>, http::Request<hyper::Body>) -> impl futures::Future<Output = Result<response::GatewayResponse, GatewayError>> {put_handler::<T>}`
    |                             |
    |                             required by a bound introduced by this call
    |
    = help: the trait `Handler<T, ReqBody>` is implemented for `Layered<S, T>`
note: required by a bound in `axum::routing::put`
   --> /home/fabrice/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.5.17/src/routing/method_routing.rs:402:1
    |
402 | top_level_handler_fn!(put, PUT);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `axum::routing::put`
    = note: this error originates in the macro `top_level_handler_fn` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Thanks!

### CODE REVIEW

**Comment by Arqu** [comment]
@fabricedesre Just a quick FYI - writeable gateways are a whole separate can of worms and currently intentionally left out.

Regarding the issue at hand, axum is a bit ugly on this front. [axum-macros](https://lib.rs/crates/axum-macros) is useful to help debug these nonsensical errors but unfortunately doesn't work with generic functions. I quickly butchered the gateway piece to test it out and here's what I get back:
```
Compiling iroh-gateway v0.1.0 (/Users/arqu/dev/n0/iroh/iroh-gateway)
warning: unused import: `BodyStream`
 --> iroh-gateway/src/handlers.rs:5:39
  |
5 |     extract::{Extension, Path, Query, BodyStream},
  |                                       ^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `futures::StreamExt`
   --> iroh-gateway/src/handlers.rs:265:9
    |
265 |     use futures::StreamExt;
    |         ^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `fn(axum::extract::Path<HashMap<std::string::String, std::string::String>>, Query<GetParams>, Method, http::Request<hyper::Body>, http::HeaderMap) -> impl futures::Future<Output = Result<response::GatewayResponse, GatewayError>> {handlers::put_handler}: Handler<_, _>` is not satisfied
   --> iroh-gateway/src/handlers.rs:72:33
    |
72  |         .route("/:scheme/", put(put_handler))
    |                             --- ^^^^^^^^^^^ the trait `Handler<_, _>` is not implemented for `fn(axum::extract::Path<HashMap<std::string::String, std::string::String>>, Query<GetParams>, Method, http::Request<hyper::Body>, http::HeaderMap) -> impl futures::Future<Output = Result<response::GatewayResponse, GatewayError>> {handlers::put_handler}`
    |                             |
    |                             required by a bound introduced by this call
    |
    = help: the trait `Handler<T, ReqBody>` is implemented for `Layered<S, T>`
note: required by a bound in `axum::routing::put`
   --> /Users/arqu/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.5.16/src/routing/method_routing.rs:400:1
    |
400 | top_level_handler_fn!(put, PUT);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `axum::routing::put`
    = note: this error originates in the macro `top_level_handler_fn` (in Nightly builds, run with -Z macro-backtrace for more info)

error: future cannot be sent between threads safely
   --> iroh-gateway/src/handlers.rs:257:1
    |
257 | pub async fn put_handler(
    | ^^^ future returned by `put_handler` is not `Send`
    |
    = help: the trait `std::marker::Send` is not implemented for `(dyn tokio::io::AsyncRead + 'static)`
note: future is not `Send` as this value is used across an await
   --> iroh-gateway/src/handlers.rs:243:25
    |
240 |     let mut file_builder = FileBuilder::new();
    |         ---------------- has type `FileBuilder` which is not `Send`
...
243 |     file_builder.build().await
    |                         ^^^^^^ await occurs here, with `mut file_builder` maybe used later
244 | }
    | - `mut file_builder` is later dropped here
note: required by a bound in `__axum_macros_check_put_handler_future::check`
   --> iroh-gateway/src/handlers.rs:257:1
    |
257 | pub async fn put_handler(
    | ^^^ required by this bound in `__axum_macros_check_put_handler_future::check`

For more information about this error, try `rustc --explain E0277`.
warning: `iroh-gateway` (lib) generated 2 warnings
error: could not compile `iroh-gateway` due to 2 previous errors; 2 warnings emitted
```

TLDR; the handler func needs to be `Send`, `FileBuilder` is not `+ Send`

Hope this helps.

**Comment by fabricedesre** [comment]
Thanks @Arqu ! Indeed that makes things not that straightforward...

File: iroh-gateway/src/handlers.rs [hunk 1]
```diff
@@ -225,6 +238,54 @@ pub async fn get_handler<T: ContentLoader + std::marker::Unpin>(
...
+
+    let cid = state
+        .client
+        .resolver
+        .loader
+        .store_file(reader)
+        .await
+        .map_err(into_gateway)?;
+    let location = format!("ipfs://{}", cid);
+
```
Comment (ramfox) [inline iroh-gateway/src/handlers.rs:280]: after you get the root CID here, you need to call `state.client.try_p2p()?.start_providing().await?`
to actually provide the content to the dht

**Comment by fabricedesre** [comment]
Can I get some feedback from the team on that? This is providing the bare minimum support to upload data and provide it. It should be interoperable with https://github.com/AgregoreWeb/agregore-ipfs-daemon/blob/main/spec.md#post-ipfs

The feature is only enabled when the gateway configuration flag `writeable` is set to true, which is not the case by default.

File: iroh-resolver/src/content_loader.rs [hunk 2]
```diff
@@ -24,6 +24,13 @@ pub trait ContentLoader: Sync + Send + std::fmt::Debug + Clone + 'static {
     async fn stop_session(&self, ctx: ContextId) -> Result<()>;
     /// Checks if the given cid is present in the local storage.
     async fn has_cid(&self, cid: &Cid) -> Result<bool>;
+    /// Store some content
+    async fn store_file<T: tokio::io::AsyncRead + 'static + std::marker::Send>(
+        &self,
+        _content: T,
+    ) -> Result<cid::Cid, anyhow::Error> {
+        unimplemented!()
```
Comment (Arqu) [inline iroh-resolver/src/content_loader.rs:32]: This is still unimplemented. Should we allow this?
On a separate note, feels wrong to be placed in the content loader. Unsure if the content loader is evolving to be more (and whether it should) or if we should move this out and have it be handled by the gateway client and not the resolver.

File: iroh-resolver/src/unixfs_builder.rs [hunk 3]
```diff
@@ -168,6 +170,8 @@ pub struct File {
     chunker: Chunker,
 }
 
+unsafe impl Send for File {}
```
Comment (Arqu) [inline iroh-resolver/src/unixfs_builder.rs:173]: Not thrilled about unsafe code. @dignifiedquire what's our general policy on that front?

**Review by Arqu — COMMENTED** [review]
I've reviewed this mostly to keep the conversation going. 
@b5 what's our roadmap/product stance on the writable gateway story?

File: iroh-resolver/src/chunker.rs [hunk 4]
```diff
@@ -35,6 +35,8 @@ pub enum ChunkerStream<'a> {
     Rabin(LocalBoxStream<'a, io::Result<Bytes>>),
 }
 
+unsafe impl<'a> Send for ChunkerStream<'a> {}
```
Comment (dignifiedquire) [inline iroh-resolver/src/chunker.rs:38]: No, this is not safe to do. The streams need to change to `BoxStream` to make this work

File: iroh-resolver/src/unixfs_builder.rs (hunk 3 above)
Comment (dignifiedquire) [reply iroh-resolver/src/unixfs_builder.rs:173]: No, this is not safe to do. Each piece of the struct needs to be made `Send`

**Review by dignifiedquire — CHANGES_REQUESTED** [review]
None of the `unsafe` usages are actually safe to use, these need reworking

**Comment by fabricedesre** [comment]
@dignifiedquire I'm not surprised the unsafe parts need to be re-done, it was mostly "compiler error" based implementation.

I'm not convinced by the overall design anyway. Like @Arqu comments in https://github.com/n0-computer/iroh/pull/499/files#r1035765976 we should settle on that question first.
